csharp/src/Ice/LocatorInfo.cs

```diff
@@ -51,123 +59,90 @@ internal void ClearCache(Reference reference)
+        ResolveIndirectReferenceAsync(
             Reference reference,
             CancellationToken cancel)
         {
             Debug.Assert(reference.IsIndirect);
-            Reference? directReference = null;
+
+            EndpointList endpoints = ImmutableArray<Endpoint>.Empty;
+            Location newLocation = ImmutableArray<string>.Empty;
             bool cached = false;
 
-            Reference? indirectReference = null;
+            Location locationToResolve;
 
             if (reference.IsWellKnown)
             {
-                Reference? resolvedWellKnownProxy;
-
-                // First, we resolve the well-known reference. The resolved reference can be direct or indirect, but
-                // cannot be well-known.
-                (resolvedWellKnownProxy, cached) = GetResolvedWellKnownProxyFromCache(reference,
+                // First, we resolve the well-known reference.
+                (endpoints, newLocation, cached) = GetResolvedWellKnownProxyFromCache(reference,
                                                                                       reference.LocatorCacheTimeout);
                 if (!cached)
                 {
-                    if (_background && resolvedWellKnownProxy != null)
+                    if (_background && (endpoints.Count > 0 || newLocation.Count > 0))
                     {
-                        // Reference is returned from the cache but TTL was reached, if backgrounds updates
-                        // are configured, we obtain a new reference to refresh the cache but use the stale
-                        // reference to not block the caller.
+                        // Entry is returned from the cache but TTL was reached, if backgrounds updates are configured,
+                        // we make a new resolution to refresh the cache but use the stale info to not block the caller.
                         _ = ResolveWellKnownProxyAsync(reference, cancel: default);
                     }
                     else
                     {
-                        resolvedWellKnownProxy =
+                        (endpoints, newLocation) =
                             await ResolveWellKnownProxyAsync(reference, cancel).ConfigureAwait(false);
                     }
                 }
 
-                if (resolvedWellKnownProxy != null)
-                {
-                    if (resolvedWellKnownProxy.IsIndirect)
-                    {
-                        indirectReference = resolvedWellKnownProxy;
-                    }
-                    else
-                    {
-                        directReference = resolvedWellKnownProxy;
-                    }
-                }
+                locationToResolve = endpoints.Count == 0 ? newLocation : ImmutableArray<string>.Empty;
             }
             else
             {
-                indirectReference = reference;
+                locationToResolve = reference.Location;
```

Comment:
can we get rid of this else and just make `reference.Location` the initial value of `locationToResolve`?

Reply:
Yes, fixed.